packages/backend/server/src/plugins/copilot/session.ts

```diff
@@ -0,0 +1,203 @@
+import { randomUUID } from 'node:crypto';
+
+import { Injectable, Logger } from '@nestjs/common';
+import { PrismaClient } from '@prisma/client';
+
+import { ChatPrompt, PromptService } from './prompt';
+import {
+  ChatMessage,
+  ChatMessageSchema,
+  PromptMessage,
+  PromptParams,
+} from './types';
+
+export interface ChatSessionOptions {
+  userId: string;
+  workspaceId: string;
+  docId: string;
+  promptName: string;
+}
+
+export interface ChatSessionState
+  extends Omit<ChatSessionOptions, 'promptName'> {
+  // connect ids
+  sessionId: string;
+  // states
+  prompt: ChatPrompt;
+  messages: ChatMessage[];
+}
+
+export class ChatSession implements AsyncDisposable {
+  constructor(
+    private readonly state: ChatSessionState,
+    private readonly dispose?: (state: ChatSessionState) => Promise<void>,
+    private readonly maxTokenSize = 3840
+  ) {}
+
+  get model() {
+    return this.state.prompt.model;
+  }
+
+  push(message: ChatMessage) {
+    this.state.messages.push(message);
+  }
+
+  pop() {
+    this.state.messages.pop();
+  }
+
+  private takeMessages(): ChatMessage[] {
+    if (this.state.prompt.action) {
+      const messages = this.state.messages;
+      return messages.slice(messages.length - 1);
+    }
+    const ret = [];
+    const messages = this.state.messages.slice();
+
+    let size = this.state.prompt.tokens;
+    while (messages.length) {
+      const message = messages.pop();
```

Comment:
`pop` is still a array mutation op,

```
for (let i = messages.length - 1; i >= 0; i--)
```

Reply:
`messages.slice()` has copied the array and will not change the original array
according to the previous test results, the while pop combination is slightly faster than for, and the code is clearer:

![image](https://github.com/toeverything/AFFiNE/assets/25152247/3b3acaba-a4e4-4384-83d8-5f47881e3b1e)